packages/plugin-data/src/commands/force/data/soql/query.ts

```diff
@@ -0,0 +1,224 @@
+      description: messages.getMessage('queryToExecute'),
+      longDescription: messages.getMessage('queryLongDescription'),
+    }),
+    usetoolingapi: flags.boolean({
+      char: 't',
+      required: false,
+      hidden: false,
+      default: false,
+      description: messages.getMessage('queryToolingDescription'),
+      longDescription: messages.getMessage('queryToolingLongDescription'),
+    }),
+    resultformat: flags.enum({
+      char: 'r',
+      description: messages.getMessage('resultFormatDescription'),
+      longDescription: messages.getMessage('resultFormatLongDescription'),
+      options: ['human', 'csv', 'json'],
+      default: 'human',
+    }),
+  };
+
+  /**
+   * Define display function that will produce the desired output based on flag resultformat selection
+   *
+   * @protected
+   */
+  protected static readonly result: SfdxResult = {
+    display(): void {
+      const results = asPlainObject(this.data) as DataSoqlQueryResult;
+      let reporter;
+      switch (results.resultFormat as keyof typeof FormatTypes) {
+        case 'human':
+          reporter = new HumanReporter(results, results.columns, this.ux, results.logger);
+          break;
+        case 'json':
+          reporter = new JsonReporter(results, results.columns, this.ux, results.logger);
+          break;
+        case 'csv':
+          reporter = new CsvReporter(results, results.columns, this.ux, results.logger);
+          break;
+        default:
+          throw new Error(`result format is invalid: ${results.resultFormat}`);
+      }
+      // delegate to selected reporter
+      reporter.display();
+    },
+  };
+
+  // Overrides SfdxCommand.  This is ensured since requiresUsername == true
+  protected org!: Org;
+
+  /**
+   * Command run implementation
+   *
+   * Returns either a DataSoqlQueryResult or a SfdxResult.
+   * When the user is using global '--json' flag an instance of SfdxResult si returned.
+   * This is necessary since '--json' flag reports results in the form of SfdxResult
+   * and bypasses the definition of start result. The goal is to have the output
+   * from '--json' and '--resulformat json' be the same.
+   *
+   * The DataSoqlQueryResult is necessary to communicate user selections to the reporters.
+   * The 'this' object available during display() function does not include user input to
+   * the command, which are necessary for reporter selection.
+   *
+   */
+  public async run(): Promise<DataSoqlQueryResult | SfdxResult> {
+    try {
+      if (this.flags.resultformat !== 'json') this.ux.startSpinner(messages.getMessage('queryRunningMessage'));
+      const query = new SoqlQuery();
+      const queryResult: SoqlQueryResult = await query.runSoqlQuery(
+        this.flags.usetoolingapi ? this.org.getConnection().tooling : this.org.getConnection(),
+        this.flags.query,
+        this.logger
+      );
+      const results = {
+        ...queryResult,
+        resultFormat: this.flags.resultformat,
```

Comment:
I don't see any of these in the old json output. Did you just decide to add them?

Reply:
@amphro The current version of the query command emitted user output during the execution of the command. I moved output to the override of the `display()` function. Currently the `this` reference made available to `display()` does not include the command parameters, etc. made by the user. I am passing along the needed data to drive selected output format.

Reply:
Ya, I just don't know if this is the best way to do it, because this becomes part of the output of `--json` right? For example, adding logger on the result will put the whole logger object in the json that is output from the terminal? Seems like we don't want that.

Reply:
If it doesn't output that information, then I guess it is fine. It seems like we are trying to fit a square peg in a round hole with the `display` function though. Maybe it would have been better just to call `reporter.display` in the run method instead...

Reply:
Or, I think you can also just say 

```
this.result = result;
return queryResult;
```

rather than having to have the normalizeJson method. This also prevents having to have two return types from the run method which will be confusing when tooling vendors consume these types which should alway mirror what is returned from the run method.